src/msg/async/ProtocolV2.cc

```diff
@@ -2607,7 +2613,9 @@ CtPtr ProtocolV2::handle_existing_connection(const AsyncConnectionRef& existing)
     ldout(cct, 1) << __func__ << " found previous session existing=" << existing
                   << ", peer must have reseted." << dendl;
     if (connection->policy.resetcheck) {
-      exproto->reset_session();
+      existing->center->submit_to(existing->center->get_id(), [exproto] {
+        exproto->reset_session();
+      });
```

Comment:
I am mostly ignorant of how the event center thing works.. is this submit_to code executed synchronously?  I'm worried that if it's fully async this will introduce a race?

Reply:
`submit_to()` in this particular appliance (where `nowait` parameter has the default value) is actually synchronous:
```cpp
  template <typename func>
  void submit_to(int i, func &&f, bool nowait = false) {
    ceph_assert(i < MAX_EVENTCENTER && global_centers);
    EventCenter *c = global_centers->centers[i];
    ceph_assert(c);
    if (nowait) {
      C_submit_event<func> *event = new C_submit_event<func>(std::move(f), true);
      c->dispatch_event_external(event);
    } else if (c->in_thread()) {
      f();
      return;
    } else {
      C_submit_event<func> event(std::move(f), false);
      c->dispatch_event_external(&event);
      event.wait();
    }
  };
```

I'm not worried about a deadlock. My reasoning is:
1. in order to acquire `AsyncConnection::write_lock`, `AsyncConnection::lock` must be locked first.
2. `reset_session()` (possibly in different thread) acquires `write_lock` very early while `lock` has been taken at the beginning of the `handle_existing_connection()` AND the method itself doesn't get `write_lock`:
```cpp
ProtocolV2::handle_existing_connection(const AsyncConnectionRef& existing) {
  ldout(cct, 20) << __func__ << " existing=" << existing << dendl;

  std::lock_guard<std::mutex> l(existing->lock);
```